versioned_docs/version-2020fa/lecture8.md

```diff
@@ -3,4 +3,561 @@ id: lecture8
+  useEffect(() => {
+    fetch(`${API}`)
+      .then((response) => response.json())
+      .then((d) => setData(d));
+  });
+  // ... other methods, return, etc
+};
+```
+
+Here are the important parts:
+
+- We have `data` in our component state. Later, we call `setData` on what the
+  backend sends to us.
+- We declare a type definition for the `useState` call so that TS knows the type
+  of data we're working with. (It can't infer type from an empty array!)
+- The `useEffect` hook is used to subscribe to new data.
+- `fetch(...)` is called on an API link, followed by `.then(...)` calls that
+  format the response into json and then `setData` to the response. The
+  `.then()` calls exist because `fetch()` returns a `Promise` (this is explained
+  below)
+
+Let's take a deeper look at `fetch`!
+
+## `fetch`
+
+"stop trying to make fetch happen 👧"
+
+`fetch(resource, [init])` is a native browser function for making web requests.
+
+Its params are:
+
+- `resource`: URL of the site you are fetching from
+- `init`: optional object containing any custom settings you want to apply to
+  the request.
+
+```typescript
+    // your init object might look like this
+    {
+      // HTTP request method
+       method: 'GET', // | 'POST' | 'PUT' | 'DELETE' | etc
+       // Any request headers you want to add
+       headers: {
+         'content-type': 'application/json'
+       },
+       // Request body (remember to stringify!)
+       body: JSON.stringify(requestBody)
+       // ... other settings
+    }
+```
+
+- For more on the init object, refer to
+  [this link](https://developer.mozilla.org/en-US/docs/Web/API/WindowOrWorkerGlobalScope/fetch)!
+
+**IMPORTANT: fetch() returns a PROMISE!**
+
+## Promises
+
+Operations like web requests don't complete instantly! You want to do other stuff
+while the operation is still going on.
+
+Promises represent the **eventual** completion (or failure) of an async operation.
+
+Promises are in one of three possible states:
+
+- `pending`: initial state; neither fulfilled nor rejected
+- `fulfilled`: operation completed successfully
+- `rejected`: operation failed
+
+### `.then()`
+
+`.then()` is a function on Promises that return a promise.
+
+```typescript
+p.then(onFulfilled[, onRejected])
+```
+
+Let's break this down!
+
+- `p` is a Promise.
+- `onFulfilled` is the callback function that is run when `p` is fulfilled
+- `onRejected` (OPTIONAL) is the callback for when `p` is rejected
+
+```typescript
+p.then(
+  (value) => {
+    // fulfillment
+  },
+  (reason) => {
+    // rejection
+  }
+);
+```
+
+Let's talk about types! `p` in this case might be `fetch()`, which returns the
+type `Promise<Response>`. If so, then `value` would have the type `Response`.
+Then, if the fulfillment function (which takes in `value`) returns type
+`string`, then the entire expression would be type `Promise<String>`.
+
+If you've taken CS 3110 or done some function programming/category theory, this
```

Comment:
"functional programming outside of this class" instead of "function programming"